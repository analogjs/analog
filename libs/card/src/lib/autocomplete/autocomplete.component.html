<mat-form-field>
  <mat-label>Test runners</mat-label>
  <input
    matInput
    type="text"
    aria-label="Test runners"
    placeholder="Pick one"
    [matAutocomplete]="autocomplete"
    [(ngModel)]="value"
  />
  <mat-autocomplete #autocomplete="matAutocomplete" autoActiveFirstOption>
    @for (option of filteredOptions(); track option) {
      <mat-option [value]="option.value">{{ option.label }}</mat-option>
    }
  </mat-autocomplete>
</mat-form-field>
