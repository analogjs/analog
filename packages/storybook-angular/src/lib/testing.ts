import type { AngularRenderer } from '@storybook/angular';
import { setProjectAnnotations as originalSetProjectAnnotations } from '@storybook/angular/dist/client/index.mjs';
import {
  NamedOrDefaultProjectAnnotations,
  NormalizedProjectAnnotations,
  RenderContext,
} from 'storybook/internal/types';
import * as configAnnotations from '@storybook/angular/dist/client/config.mjs';

export const render = configAnnotations.render;

export async function renderToCanvas(
  context: RenderContext<AngularRenderer>,
  element: HTMLElement,
) {
  element.id = context.id;
  await configAnnotations.renderToCanvas(context, element);
}

const renderAnnotations = {
  render,
  renderToCanvas,
};

export function setProjectAnnotations(
  projectAnnotations:
    | NamedOrDefaultProjectAnnotations<any>
    | NamedOrDefaultProjectAnnotations<any>[],
): NormalizedProjectAnnotations<AngularRenderer> {
  return originalSetProjectAnnotations([
    renderAnnotations,
    ...(Array.isArray(projectAnnotations)
      ? projectAnnotations
      : [projectAnnotations]),
  ]) as NormalizedProjectAnnotations<AngularRenderer>;
}
