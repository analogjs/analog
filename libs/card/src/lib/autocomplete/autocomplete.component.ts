import {
  ChangeDetectionStrategy,
  Component,
  computed,
  model,
} from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { MatAutocompleteModule } from '@angular/material/autocomplete';
import { MatInputModule } from '@angular/material/input';

@Component({
  standalone: true,
  selector: 'app-autocomplete',
  templateUrl: './autocomplete.component.html',
  styleUrl: './autocomplete.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
  imports: [
    FormsModule,
    ReactiveFormsModule,
    MatAutocompleteModule,
    MatInputModule,
  ],
})
export class AutocompleteComponent {
  public readonly value = model.required<string>();

  protected readonly options = Object.freeze([
    {
      label: 'Jest',
      value: TestRunner.Jest,
    },
    {
      label: 'Karma',
      value: TestRunner.Karma,
    },
    {
      label: 'Vitest',
      value: TestRunner.Vitest,
    },
  ]);

  protected readonly filteredOptions = computed(() => {
    const value = this.value().toLocaleLowerCase().trim();

    return Object.freeze(
      this.options.filter((option) =>
        option.label.toLocaleLowerCase().includes(value),
      ),
    );
  });
}

export const enum TestRunner {
  Jest,
  Karma,
  Vitest,
}
