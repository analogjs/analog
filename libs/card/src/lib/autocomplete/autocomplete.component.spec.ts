import { ComponentFixture, TestBed } from '@angular/core/testing';
import { type HarnessLoader } from '@angular/cdk/testing';
import { TestbedHarnessEnvironment } from '@angular/cdk/testing/testbed';
import { MatAutocompleteHarness } from '@angular/material/autocomplete/testing';
import { provideNoopAnimations } from '@angular/platform-browser/animations';

import { AutocompleteComponent } from './autocomplete.component';

describe('GIVEN AutocompleteComponent', () => {
  let fixture: ComponentFixture<AutocompleteComponent>;
  let harnessLoader: HarnessLoader;
  let component: AutocompleteComponent;

  beforeEach(() => {
    TestBed.configureTestingModule({
      providers: [provideNoopAnimations()],
      imports: [AutocompleteComponent],
    });

    fixture = TestBed.createComponent(AutocompleteComponent);
    harnessLoader = TestbedHarnessEnvironment.loader(fixture);
    component = fixture.componentInstance;
  });

  describe('WHEN input value is empty', () => {
    beforeEach(() => {
      fixture.componentRef.setInput('value', '');
      fixture.detectChanges();
    });

    it('THEN component should be initialized', () => {
      expect(component).toBeTruthy();
    });

    it('THEN component should match snapshot', () => {
      expect(fixture.debugElement.nativeElement).toMatchSnapshot();
    });

    describe('AND input is focused', () => {
      let autocomplete: MatAutocompleteHarness;

      beforeEach(async () => {
        autocomplete = await harnessLoader.getHarness(MatAutocompleteHarness);

        await autocomplete.focus();
      });

      it('THEN component should match snapshot', () => {
        expect(fixture.debugElement.nativeElement).toMatchSnapshot();
      });

      it('THEN autocomplete should be opened', async () => {
        const actual = await autocomplete.isOpen();

        expect(actual).toBe(true);
      });

      it('THEN autocomplete should have 3 options', async () => {
        const options = await autocomplete.getOptions();
        const actual = options.length;

        expect(actual).toBe(3);
      });
    });
  });

  describe('WHEN input value is not empty', () => {
    beforeEach(() => {
      fixture.componentRef.setInput('value', 'est');
      fixture.detectChanges();
    });

    it('THEN component should be initialized', () => {
      expect(component).toBeTruthy();
    });

    it('THEN component should match snapshot', () => {
      expect(fixture.debugElement.nativeElement).toMatchSnapshot();
    });

    describe('AND input is focused', () => {
      let autocomplete: MatAutocompleteHarness;

      beforeEach(async () => {
        autocomplete = await harnessLoader.getHarness(MatAutocompleteHarness);

        await autocomplete.focus();
      });

      it('THEN component should match snapshot', () => {
        expect(fixture.debugElement.nativeElement).toMatchSnapshot();
      });

      it('THEN autocomplete should be opened', async () => {
        const actual = await autocomplete.isOpen();

        expect(actual).toBe(true);
      });

      it('THEN autocomplete should have 2 options', async () => {
        const options = await autocomplete.getOptions();
        const actual = options.length;

        expect(actual).toBe(2);
      });
    });
  });
});
