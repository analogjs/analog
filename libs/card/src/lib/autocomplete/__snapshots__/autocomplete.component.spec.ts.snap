// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GIVEN AutocompleteComponent > WHEN input value is empty > AND input is focused > THEN component should match snapshot 1`] = `
<div
  id="root2"
  ng-version="19.0.0"
>
  <mat-form-field
    class="mat-mdc-form-field ng-tns-c508571215-0 mat-mdc-form-field-type-mat-input mat-form-field-no-animations mat-form-field-appearance-fill mat-primary ng-untouched ng-pristine ng-valid mat-mdc-form-field-label-always-float mat-focused"
  >
    <!--container-->
    <div
      class="mat-mdc-text-field-wrapper mdc-text-field ng-tns-c508571215-0 mdc-text-field--filled mdc-text-field--focused"
    >
      <div
        class="mat-mdc-form-field-focus-overlay ng-tns-c508571215-0 ng-star-inserted"
      />
      <!--container-->
      <div
        class="mat-mdc-form-field-flex ng-tns-c508571215-0"
      >
        <!--container-->
        <!--container-->
        <!--container-->
        <div
          class="mat-mdc-form-field-infix ng-tns-c508571215-0"
        >
          <label
            class="mdc-floating-label mat-mdc-floating-label ng-tns-c508571215-0 ng-star-inserted mdc-floating-label--float-above"
            for="mat-input-a2"
            id="mat-mdc-form-field-label-a2"
            matformfieldfloatinglabel=""
            ng-reflect-floating="true"
            ng-reflect-monitor-resize="false"
          >
            <mat-label
              class="ng-tns-c508571215-0"
            >
              Test runners
            </mat-label>
            <!--container-->
          </label>
          <!--container-->
          <!--bindings={
  "ng-reflect-ng-template-outlet": "[object Object]"
}-->
          <!--container-->
          <input
            aria-activedescendant="mat-option-a6"
            aria-autocomplete="list"
            aria-controls="mat-autocomplete-a2"
            aria-expanded="true"
            aria-haspopup="listbox"
            aria-invalid="false"
            aria-label="Test runners"
            aria-required="false"
            autocomplete="off"
            class="mat-mdc-autocomplete-trigger mat-mdc-input-element ng-tns-c508571215-0 ng-untouched ng-pristine ng-valid mat-mdc-form-field-input-control mdc-text-field__input cdk-text-field-autofill-monitored"
            id="mat-input-a2"
            matinput=""
            ng-reflect-autocomplete="[object Object]"
            ng-reflect-model=""
            ng-reflect-placeholder="Pick one"
            ng-reflect-type="text"
            placeholder="Pick one"
            role="combobox"
            type="text"
          />
          <!--container-->
          <mat-autocomplete
            autoactivefirstoption=""
            class="mat-mdc-autocomplete ng-tns-c3964489799-1 ng-tns-c508571215-0 ng-star-inserted"
            ng-reflect-auto-active-first-option=""
          >
            <!--container-->
          </mat-autocomplete>
        </div>
        <!--container-->
        <!--container-->
      </div>
      <div
        class="mdc-line-ripple ng-tns-c508571215-0 ng-star-inserted mdc-line-ripple--active"
        matformfieldlineripple=""
      />
      <!--container-->
    </div>
    <div
      class="mat-mdc-form-field-subscript-wrapper mat-mdc-form-field-bottom-align ng-tns-c508571215-0"
    >
      <!--container-->
      <div
        class="mat-mdc-form-field-hint-wrapper ng-tns-c508571215-0 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
        style="opacity: 1; transform: translateY(0%);"
      >
        <!--container-->
        <div
          class="mat-mdc-form-field-hint-spacer ng-tns-c508571215-0"
        />
      </div>
      <!--container-->
    </div>
  </mat-form-field>
</div>
`;

exports[`GIVEN AutocompleteComponent > WHEN input value is empty > THEN component should match snapshot 1`] = `
<div
  id="root1"
  ng-version="19.0.0"
>
  <mat-form-field
    class="mat-mdc-form-field ng-tns-c508571215-0 mat-mdc-form-field-type-mat-input mat-form-field-no-animations mat-form-field-appearance-fill mat-form-field-hide-placeholder mat-primary ng-untouched ng-pristine ng-valid"
  >
    <!--container-->
    <div
      class="mat-mdc-text-field-wrapper mdc-text-field ng-tns-c508571215-0 mdc-text-field--filled"
    >
      <div
        class="mat-mdc-form-field-focus-overlay ng-tns-c508571215-0 ng-star-inserted"
      />
      <!--container-->
      <div
        class="mat-mdc-form-field-flex ng-tns-c508571215-0"
      >
        <!--container-->
        <!--container-->
        <!--container-->
        <div
          class="mat-mdc-form-field-infix ng-tns-c508571215-0"
        >
          <label
            class="mdc-floating-label mat-mdc-floating-label ng-tns-c508571215-0 ng-star-inserted"
            for="mat-input-a1"
            id="mat-mdc-form-field-label-a1"
            matformfieldfloatinglabel=""
            ng-reflect-floating="false"
            ng-reflect-monitor-resize="false"
          >
            <mat-label
              class="ng-tns-c508571215-0"
            >
              Test runners
            </mat-label>
            <!--container-->
          </label>
          <!--container-->
          <!--bindings={
  "ng-reflect-ng-template-outlet": "[object Object]"
}-->
          <!--container-->
          <input
            aria-autocomplete="list"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-invalid="false"
            aria-label="Test runners"
            aria-required="false"
            autocomplete="off"
            class="mat-mdc-autocomplete-trigger mat-mdc-input-element ng-tns-c508571215-0 ng-untouched ng-pristine ng-valid mat-mdc-form-field-input-control mdc-text-field__input cdk-text-field-autofill-monitored"
            id="mat-input-a1"
            matinput=""
            ng-reflect-autocomplete="[object Object]"
            ng-reflect-model=""
            ng-reflect-placeholder="Pick one"
            ng-reflect-type="text"
            placeholder="Pick one"
            role="combobox"
            type="text"
          />
          <!--container-->
          <mat-autocomplete
            autoactivefirstoption=""
            class="mat-mdc-autocomplete ng-tns-c3964489799-1 ng-tns-c508571215-0 ng-star-inserted"
            ng-reflect-auto-active-first-option=""
          >
            <!--container-->
          </mat-autocomplete>
        </div>
        <!--container-->
        <!--container-->
      </div>
      <div
        class="mdc-line-ripple ng-tns-c508571215-0 mdc-line-ripple--deactivating ng-star-inserted"
        matformfieldlineripple=""
      />
      <!--container-->
    </div>
    <div
      class="mat-mdc-form-field-subscript-wrapper mat-mdc-form-field-bottom-align ng-tns-c508571215-0"
    >
      <!--container-->
      <div
        class="mat-mdc-form-field-hint-wrapper ng-tns-c508571215-0 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
        style="opacity: 1; transform: translateY(0%);"
      >
        <!--container-->
        <div
          class="mat-mdc-form-field-hint-spacer ng-tns-c508571215-0"
        />
      </div>
      <!--container-->
    </div>
  </mat-form-field>
</div>
`;

exports[`GIVEN AutocompleteComponent > WHEN input value is not empty > AND input is focused > THEN component should match snapshot 1`] = `
<div
  id="root7"
  ng-version="19.0.0"
>
  <mat-form-field
    class="mat-mdc-form-field ng-tns-c508571215-0 mat-mdc-form-field-type-mat-input mat-form-field-no-animations mat-form-field-appearance-fill mat-primary ng-untouched ng-pristine ng-valid mat-mdc-form-field-label-always-float mat-focused"
  >
    <!--container-->
    <div
      class="mat-mdc-text-field-wrapper mdc-text-field ng-tns-c508571215-0 mdc-text-field--filled mdc-text-field--focused"
    >
      <div
        class="mat-mdc-form-field-focus-overlay ng-tns-c508571215-0 ng-star-inserted"
      />
      <!--container-->
      <div
        class="mat-mdc-form-field-flex ng-tns-c508571215-0"
      >
        <!--container-->
        <!--container-->
        <!--container-->
        <div
          class="mat-mdc-form-field-infix ng-tns-c508571215-0"
        >
          <label
            class="mdc-floating-label mat-mdc-floating-label ng-tns-c508571215-0 ng-star-inserted mdc-floating-label--float-above"
            for="mat-input-a7"
            id="mat-mdc-form-field-label-a7"
            matformfieldfloatinglabel=""
            ng-reflect-floating="true"
            ng-reflect-monitor-resize="false"
          >
            <mat-label
              class="ng-tns-c508571215-0"
            >
              Test runners
            </mat-label>
            <!--container-->
          </label>
          <!--container-->
          <!--bindings={
  "ng-reflect-ng-template-outlet": "[object Object]"
}-->
          <!--container-->
          <input
            aria-activedescendant="mat-option-a19"
            aria-autocomplete="list"
            aria-controls="mat-autocomplete-a7"
            aria-expanded="true"
            aria-haspopup="listbox"
            aria-invalid="false"
            aria-label="Test runners"
            aria-required="false"
            autocomplete="off"
            class="mat-mdc-autocomplete-trigger mat-mdc-input-element ng-tns-c508571215-0 ng-untouched ng-pristine ng-valid mat-mdc-form-field-input-control mdc-text-field__input cdk-text-field-autofill-monitored"
            id="mat-input-a7"
            matinput=""
            ng-reflect-autocomplete="[object Object]"
            ng-reflect-model="est"
            ng-reflect-placeholder="Pick one"
            ng-reflect-type="text"
            placeholder="Pick one"
            role="combobox"
            type="text"
          />
          <!--container-->
          <mat-autocomplete
            autoactivefirstoption=""
            class="mat-mdc-autocomplete ng-tns-c3964489799-1 ng-tns-c508571215-0 ng-star-inserted"
            ng-reflect-auto-active-first-option=""
          >
            <!--container-->
          </mat-autocomplete>
        </div>
        <!--container-->
        <!--container-->
      </div>
      <div
        class="mdc-line-ripple ng-tns-c508571215-0 ng-star-inserted mdc-line-ripple--active"
        matformfieldlineripple=""
      />
      <!--container-->
    </div>
    <div
      class="mat-mdc-form-field-subscript-wrapper mat-mdc-form-field-bottom-align ng-tns-c508571215-0"
    >
      <!--container-->
      <div
        class="mat-mdc-form-field-hint-wrapper ng-tns-c508571215-0 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
        style="opacity: 1; transform: translateY(0%);"
      >
        <!--container-->
        <div
          class="mat-mdc-form-field-hint-spacer ng-tns-c508571215-0"
        />
      </div>
      <!--container-->
    </div>
  </mat-form-field>
</div>
`;

exports[`GIVEN AutocompleteComponent > WHEN input value is not empty > THEN component should match snapshot 1`] = `
<div
  id="root6"
  ng-version="19.0.0"
>
  <mat-form-field
    class="mat-mdc-form-field ng-tns-c508571215-0 mat-mdc-form-field-type-mat-input mat-form-field-no-animations mat-form-field-appearance-fill mat-form-field-hide-placeholder mat-primary ng-untouched ng-pristine ng-valid"
  >
    <!--container-->
    <div
      class="mat-mdc-text-field-wrapper mdc-text-field ng-tns-c508571215-0 mdc-text-field--filled"
    >
      <div
        class="mat-mdc-form-field-focus-overlay ng-tns-c508571215-0 ng-star-inserted"
      />
      <!--container-->
      <div
        class="mat-mdc-form-field-flex ng-tns-c508571215-0"
      >
        <!--container-->
        <!--container-->
        <!--container-->
        <div
          class="mat-mdc-form-field-infix ng-tns-c508571215-0"
        >
          <label
            class="mdc-floating-label mat-mdc-floating-label ng-tns-c508571215-0 ng-star-inserted"
            for="mat-input-a6"
            id="mat-mdc-form-field-label-a6"
            matformfieldfloatinglabel=""
            ng-reflect-floating="false"
            ng-reflect-monitor-resize="false"
          >
            <mat-label
              class="ng-tns-c508571215-0"
            >
              Test runners
            </mat-label>
            <!--container-->
          </label>
          <!--container-->
          <!--bindings={
  "ng-reflect-ng-template-outlet": "[object Object]"
}-->
          <!--container-->
          <input
            aria-autocomplete="list"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-invalid="false"
            aria-label="Test runners"
            aria-required="false"
            autocomplete="off"
            class="mat-mdc-autocomplete-trigger mat-mdc-input-element ng-tns-c508571215-0 ng-untouched ng-pristine ng-valid mat-mdc-form-field-input-control mdc-text-field__input cdk-text-field-autofill-monitored"
            id="mat-input-a6"
            matinput=""
            ng-reflect-autocomplete="[object Object]"
            ng-reflect-model="est"
            ng-reflect-placeholder="Pick one"
            ng-reflect-type="text"
            placeholder="Pick one"
            role="combobox"
            type="text"
          />
          <!--container-->
          <mat-autocomplete
            autoactivefirstoption=""
            class="mat-mdc-autocomplete ng-tns-c3964489799-1 ng-tns-c508571215-0 ng-star-inserted"
            ng-reflect-auto-active-first-option=""
          >
            <!--container-->
          </mat-autocomplete>
        </div>
        <!--container-->
        <!--container-->
      </div>
      <div
        class="mdc-line-ripple ng-tns-c508571215-0 mdc-line-ripple--deactivating ng-star-inserted"
        matformfieldlineripple=""
      />
      <!--container-->
    </div>
    <div
      class="mat-mdc-form-field-subscript-wrapper mat-mdc-form-field-bottom-align ng-tns-c508571215-0"
    >
      <!--container-->
      <div
        class="mat-mdc-form-field-hint-wrapper ng-tns-c508571215-0 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
        style="opacity: 1; transform: translateY(0%);"
      >
        <!--container-->
        <div
          class="mat-mdc-form-field-hint-spacer ng-tns-c508571215-0"
        />
      </div>
      <!--container-->
    </div>
  </mat-form-field>
</div>
`;
